crates/sui-graphql-rpc/src/types/dynamic_field.rs

```diff
@@ -172,66 +160,144 @@ impl DynamicField {
+
+        let checkpoint_viewed_at =
+            match validate_cursor_consistency(checkpoint_viewed_at, page.after(), page.before()) {
+                Ok(checkpoint_viewed_at) => checkpoint_viewed_at,
+                Err(_) => return Ok(conn),
+            };
+
+        let response = db
+            .execute_repeatable(move |conn| {
+                let (lhs, mut rhs) = Checkpoint::available_range(conn)?;
+
+                if let Some(checkpoint_viewed_at) = checkpoint_viewed_at {
+                    if checkpoint_viewed_at < lhs || rhs < checkpoint_viewed_at {
+                        return Ok::<_, diesel::result::Error>(None);
+                    }
+                    rhs = checkpoint_viewed_at;
+                }
+
+                let result = page.paginate_raw_query::<StoredHistoryObject, _>(
+                    conn,
+                    move |element| {
+                        element.map(|obj| {
+                            Cursor::new(HistoricalObjectCursor::new(
+                                obj.object_id.clone(),
+                                rhs as u64,
+                            ))
+                        })
+                    },
+                    move || {
+                        let final_ = Object::build_raw_consistent_query(
+                            |query| Self::filter(query, parent),
+                            lhs as i64,
+                            rhs as i64,
+                        );
+
+                        filter!(final_, "newer.object_version IS NULL")
+                    },
+                )?;
+
+                Ok(Some((result.0, result.1, result.2, rhs)))
             })
             .await?;
 
-        let mut conn = Connection::new(prev, next);
+        let Some((prev, next, results, rhs)) = response else {
+            return Ok(conn);
+        };
+
+        conn.has_previous_page = prev;
+        conn.has_next_page = next;
 
         for stored in results {
-            let cursor = stored.cursor().encode_cursor();
-            let field = DynamicField::try_from(stored)?;
-            conn.edges.push(Edge::new(cursor, field));
+            let cursor = Cursor::new(HistoricalObjectCursor::new(
+                stored.object_id.clone(),
+                rhs as u64,
+            ))
+            .encode_cursor();
+
+            let object = Object::try_from_stored_history_object(stored, Some(rhs as u64))?;
+
+            let move_ = MoveObject::try_from(&object).map_err(|_| {
+                Error::Internal(format!(
+                    "Failed to deserialize as Move object: {}",
+                    object.address
+                ))
+            })?;
+
+            let dynamic_field = DynamicField::try_from(move_)?;
+
+            conn.edges.push(Edge::new(cursor, dynamic_field));
         }
 
         Ok(conn)
     }
+
+    pub(crate) fn filter(query: RawQuery, parent: SuiAddress) -> RawQuery {
```

Comment:
So, taking the idea that I mentioned about bounding the parent version, here you will need to supply a parent version as well and bound the `object_version` from above with that, which also means you can't always apply the `newer.object_version IS NULL` test.

You also can't always apply the version bound because in the case of querying dynamic fields on an `Owner` which we are using as a stand-in for a wrapped object, we can't readily get the version of the object.  There I would say it's OK to use the latest version in the checkpoint.

Reply:
I think that's fine since only the Object query has the ObjectFilter. Coins and DynamicFields don't have the same filtering conditions. Thus the presence of parent_version should serve as a replacement for the filter.object_keys test

Reply:
Yeah, I'd be open to adding an optional `parentVersion` bound to the dynamic field related query functions in `Owner` especially now they aren't part of the `IOwner` interface.